### DISCUSSION TITLE
Why GraphGroup is on {GraphId} and SharePoint and Teams on {Site.GroupId}

### DISCUSSION THREAD

**Discussion by czullu** [pnp/pnpcore#440] [Q&A] [answered]
i try to understand why i need to create a new pnpcontext with CreateAsync for the GraphGroup while SharePoint and Teams can use the same?

Is there a specific reason for this separation or could
internal static async Task ConfigureForGroup(PnPContext context, Guid groupId) from PnPContextFactory be public as well?

**Comment by jansenbe** [comment, marked as answer]
@czullu : not sure if I fully understand the ask. The CreateAsync overload that takes a groupid is meant for scenarios where you don't know the SharePoint URL (e.g. a Teams app). If you have the SharePoint site URL it's best to use on overload passing in the site URL.

**Reply by czullu** [reply]
maybe is misunderstand something - but here simplified what i have been trying
```
 using (var context = await pnpContextFactory.CreateAsync(new Uri("https://tenantname.sharepoint.com/sites/sitename")))
{
                    await context.Web.EnsurePropertiesAsync(w => w.Title, w=>w.Id);
                    await context.Site.EnsurePropertiesAsync(s => s.Id, s => s.GroupId);
                    await context.Group.EnsurePropertiesAsync(g => g.Visibility);
}
```
i try to load here in this example the visibility of the group as i'm not having that info somewhere else (same for mailsettings).

Loading Web and Site Data works, but loading the Group visibility throws Error:
PnP.Core.ClientException: 'Unresolved tokens found in API call groups/{GraphId}?$select=id%2cvisibility'
which makes sense as the GraphId in Entity never gets set and i found no other way then taking the GroupId returned from SharePoint and create a new Context like this:
 ```
using (var grpContext = await pnpContextFactory.CreateAsync(new Guid(context.Site.GroupId)))
{
await grpContext.Group.EnsurePropertiesAsync(g => g.Visibility);
}
```
which works - but i was wondering why i have to go that extra step since from looking at the code i think it would be ok to have the possibility to set the graphid to groupId from SP same way as in ConfigureForGroup.
```
(context.Group as GraphGroup).Metadata.Add(PnPConstants.MetaDataGraphId, groupId.ToString());
```
but since all is internal it can not be done outside the SDK (without reflection) and i was trying to find a reason for it.

Did i overlook something or get it wrong?

**Reply by czullu** [reply]
link to associated pull request https://github.com/pnp/pnpcore/pull/442

**Reply by jansenbe** [reply]
I now better understand what you need, this indeed should get fixed. I'll have a look

**Reply by jansenbe** [reply]
https://github.com/pnp/pnpcore/commit/8442e95dcb3f3b3d2d39b5b513ddebbc5c984a2f fixes this
